### PR TITLE
Add ConViT implementation

### PR DESCRIPTION
**Pull request by aditya-agrawal-30502**

### CODE REVIEW

**Comment by aditya-agrawal-30502** [comment]
#30

File: vformer/attention/__init__.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from .cross import CrossAttention
+from .gatedpositional import GatedPositionalSelfAttention
```
Comment (NeelayS) [inline vformer/attention/__init__.py:2]: Could you please change the file name to `gated_positional.py`?

File: vformer/attention/gatedpositional.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    def __init__(self, dim, num_heads=8, head_dim=64, p_dropout=0):
+        super().__init__(dim, num_heads, head_dim, p_dropout)
+
+        self.gate = nn.Parameter(torch.ones([1, num_heads, 1, 1]))
+        self.pos = nn.Linear(3, num_heads)
+        self.n_prev = 0
+
+    def rel_embedding(self, n):
+        l = int(n ** 0.5)
```
Comment (NeelayS) [inline vformer/attention/gatedpositional.py:35]:
```suggestion

        l = int(n ** 0.5)
```

File: vformer/encoder/conv.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            depth,
+            num_heads,
+            head_dim,
+            mlp_dim,
+            p_dropout,
+            attn_dropout,
+            drop_path_rate,
+        )
+        self.encoder = nn.ModuleList([])
+        for _ in range(depth):
```
Comment (NeelayS) [inline vformer/encoder/conv.py:57]:
```suggestion

        for _ in range(depth):
```

File: vformer/models/classification/__init__.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 from .cct import CCT
+from .conv import ConViT
```
Comment (NeelayS) [inline vformer/models/classification/__init__.py:2]: Could you please change the name of the file to `convit.py`?

File: vformer/models/classification/conv.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import torch
+from einops import repeat
+
+from ...encoder import ConvitEncoder
```
Comment (NeelayS) [inline vformer/models/classification/conv.py:4]:
```suggestion
from ...encoder import ConViTEncoder
```

File: vformer/models/classification/conv.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            depth_sa,
+            attn_heads_sa,
+            encoder_mlp_dim,
+            in_channels,
+            decoder_config,
+            pool,
+            p_dropout_encoder,
+            p_dropout_embedding,
+        )
+        self.encoder_gpsa = ConvitEncoder(
```
Comment (NeelayS) [inline vformer/models/classification/conv.py:82]:
```suggestion 

        self.encoder_gpsa = ConViTEncoder(
```

File: vformer/encoder/__init__.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+from .conv import ConvitEncoder
```
Comment (NeelayS) [inline vformer/encoder/__init__.py:1]: Could you please change the name of the file to `convit.py`?

**Review by NeelayS — CHANGES_REQUESTED** [review]

**Pull request merged**